pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -650,134 +650,150 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 
 	// Number of running containers to keep.
 	keepCount := 0
-	if pod.Spec.VirtualMachine == nil {
-
-		// check the status of containers.
-		for idx, container := range pod.Spec.Containers {
-			containerStatus := podStatus.FindContainerStatusByName(container.Name)
-
-			// Call internal container post-stop lifecycle hook for any non-running container so that any
-			// allocated cpus are released immediately. If the container is restarted, cpus will be re-allocated
-			// to it.
-			if containerStatus != nil && containerStatus.State != kubecontainer.ContainerStateRunning {
-				if err := m.internalLifecycle.PostStopContainer(containerStatus.ID.ID); err != nil {
-					klog.Errorf("internal container post-stop lifecycle hook failed for container %v in pod %v with error %v",
-						container.Name, pod.Name, err)
-				}
+
+	// check the status of containers.
+	for idx, container := range pod.Spec.Containers {
+		containerStatus := podStatus.FindContainerStatusByName(container.Name)
+
+		// Call internal container post-stop lifecycle hook for any non-running container so that any
+		// allocated cpus are released immediately. If the container is restarted, cpus will be re-allocated
+		// to it.
+		if containerStatus != nil && containerStatus.State != kubecontainer.ContainerStateRunning {
+			if err := m.internalLifecycle.PostStopContainer(containerStatus.ID.ID); err != nil {
+				klog.Errorf("internal container post-stop lifecycle hook failed for container %v in pod %v with error %v",
+					container.Name, pod.Name, err)
 			}
+		}
 
-			// If container does not exist, or is not running, check whether we
-			// need to restart it.
-			if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
-				if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
-					message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
-					klog.V(3).Infof(message)
-					changes.ContainersToStart = append(changes.ContainersToStart, idx)
-					if containerStatus != nil && containerStatus.State == kubecontainer.ContainerStateUnknown {
-						// If container is in unknown state, we don't know whether it
-						// is actually running or not, always try killing it before
-						// restart to avoid having 2 running instances of the same container.
-						changes.ContainersToKill[containerStatus.ID] = containerToKillInfo{
-							name:      containerStatus.Name,
-							container: &pod.Spec.Containers[idx],
-							message:   fmt.Sprintf("Container is in %q state, try killing it before restart", containerStatus.State),
-						}
+		// If container does not exist, or is not running, check whether we
+		// need to restart it.
```

Comment:
this logic will need to handle vm cases. i.e. in here the check for workload type in line 653 on left side should NOT be ignored or the VirtualMachineStatus is needed to check.

The prelimary testing with this PR proves that the RemoveContianer and restart VM.

Reply:
Is RestartPolicy not applicable to VMs? I'm not sure why we need to check VirtualMachineStatus here at all. With the translation, we should apply this check to both containers and VMs unless there is some reason I missed.

Reply:
makes a sense. the restart of VM is caused by other issues as we just investigated. so i just close this one. in term of RestartPolicy enforcement or application to vm types. it can be tracked in the design doc. or applied in a case by case logic.